src/dmd/lexer.d

```diff
@@ -202,6 +202,129 @@ unittest
     }
 }
 
+
+/**************
+ * Pool allocator for Token objects, used by the Lexer.
+ * Tokens are never freed so Token pointers will never change
+ */
+private struct TokenPool{
+private:
+    enum PoolSize = 1024; //number of tokens per pool
+    enum DefaultPoolListSize = 64; //number of pools to allocate by default
+    Token[][] pools;
+    size_t poolsAllocated = 0;
+
+    nothrow:
+
+    ///Token pool looks like a big array
+    public ref Token getByIndex(size_t index){
+        if (index >= PoolSize*poolsAllocated)
+        {
+            allocatePools(index + 1); //need space for index 0
+        }
+        return pools[index / PoolSize][index % PoolSize];
+
+    }
+
+    private void allocatePools(size_t tokensNeeded){
+        const poolsNeeded = (tokensNeeded + PoolSize - 1)/PoolSize;
+        assert(poolsNeeded == poolsAllocated + 1); //don't alloc more than 1 pool at a time
+        if (poolsNeeded > pools.length)
+        {
+            const newLength = pools.length > 0 ? 2*pools.length : DefaultPoolListSize;
+            Token[][] newPoolList = (cast(Token[]*)mem.xmalloc((Token[]).sizeof *newLength))[0 .. newLength];
+
+            if(pools.length > 0)
+                newPoolList[0 .. pools.length] = pools[];
+
+            if (pools.ptr)
+                mem.xfree(pools.ptr);
+
+            pools = newPoolList;
+        }
+        pools[poolsAllocated++] = new Token[PoolSize]; //will never be freed, so use GC mem
+    }
+}
+
+
+/***************
+ * Represents a range of tokens lazily scanned from a lexer
+ * should be created with the Lexer makeRangeFromHere method
+ */
+struct TokenRange
+{
+private:
+    Lexer lex; //where are these tokens coming from?
+    size_t start_, stop_;
+
+public:
+    nothrow:
+    /// returns the first token in the range
```

Comment:
Since these are standard functions, there's no need to document them unless they do something non-standard.